ynr/apps/sopn_parsing/management/commands/sopn_parsing_parse_tables.py

```diff
@@ -0,0 +1,185 @@
+        return description
+
+    def get_description(self, description, sopn):
+        description = self.clean_description(description)
+
+        if not description:
+            return None
+        if description in INDEPENDENT_VALUES:
+            return None
+
+        try:
+            description_model = PartyDescription.objects.get(
+                description=description
+            )
+        except PartyDescription.DoesNotExist:
+            try:
+                description_model = PartyDescription.objects.filter(
+                    description__startswith=description,
+                    party__register=sopn.sopn.post_election.post.party_set.slug,
+                ).first()
+            except PartyDescription.DoesNotExist:
+                return None
+        return description_model
+
+    def get_party(self, description_model, description, sopn):
+        if description_model:
+            return description_model.party
+
+        party_name = self.clean_description(description)
+        qs = Party.objects.register(sopn.sopn.post_election.post.party_set.slug)
+        if not party_name:
+            return Party.objects.get(ec_id="ynmp-party:2")
+
+        try:
+            party_obj = qs.get(name=party_name)
+        except Party.DoesNotExist:
+            party_obj = None
+
+        if not party_obj:
+            raise ValueError("Unknown party")
+        return party_obj
+
+    def handle(self, *args, **options):
+
+        qs = ParsedSOPN.objects.filter(parsed_data=None, status="unparsed")
+        for sopn in qs:
+            data = sopn.as_pandas
+
+            cell_counts = [
+                len(self.merge_row_cells(c)) for c in self.iter_rows(data)
+            ]
+
+            header_found = False
+            avg_row = sum(cell_counts) / float(len(cell_counts))
+            for i, row in enumerate(self.iter_rows(data)):
+                if not header_found:
+                    if self.looks_like_header(row, avg_row):
+                        data.columns = row
+                        data = data.drop(data.index[i])
+                        header_found = True
+                    else:
+                        try:
+                            data = data.drop(data.index[i])
+                        except IndexError:
+                            break
+            if not header_found:
+                # Don't try to parse if we don't think we know the header
+                continue
+            # We're now in a position where we think we have the table we want
+            # with the columns set and other header rows removed.
+            # Time to parse it in to names and parties
+            try:
+                ballot_data = self.parse_table(sopn, data)
+            except ValueError:
+                # Something went wrong. This will happen a lot. let's move on
+                continue
+
+            if ballot_data:
+                RawPeople.objects.update_or_create(
```

Comment:
Now that I think of it, maybe that it hasn't got a lock suggestion might be the right check?

Reply:
Yes good catch. The query is now 

```py
qs = ParsedSOPN.objects.filter(
            parsed_data=None,
            status="unparsed",
            sopn__post_election__rawpeople=None,
            sopn__post_election__candidates_locked=False,
            sopn__post_election__suggestedpostlock=None,
        )
```

I think that catches every case. Note I'm checking for both locked and suggested locks, as we clean up suggestions after locking.

Reply:
I'm slightly trying to second guess your intention here, but doesn't this now create the opposite problem, which is that once a `RawPeople` exists, we can't ever re-parse. Maybe that's what you want.

If that is the intention though, can't we just simplify this. If you do `ParsedSOPN.objects.filter(sopn__post_election__rawpeople=None)` and then do `RawPeople.objects.update_or_create()`, can't you just do a safer `RawPeople.objects.create()` ?

..but looking at the comment

> I don't want to have to re-parse everything in order to run an updated table parser

and the fact one of your commands has a `--reparse` option, it seems like this is a process you're imagining running more than once..

I think it would help to understand more about your _intent_ rather than just the implementation.

Reply:
You're right, I did this in isolation of the args to the commands – I'm reviewing them, and in this case it would mean making some of the filters optional. 

I'm starting to think that it's more useful to think of all of this being run as a celery task where we parse the pages, extract and parse tables in one go. It would then be fired off after document creation, rather than as a command. 

I'm going to keep noodling on this.

The other case is: we've found a way to fix a bug in parsing and we want to reparse all documents that haven't been processed yet. That's what I was thinking of with the ``--reparse` option.

Reply:
Yeah it does feel like at the moment this is a bit of a sledgehammer - the only approach we've got is to batch-process every available SOPN.

Thinking of it as a post-process for SOPN upload might be helpful, but if its a background process after SOPN upload, you probably also want to think about how to message to a user "give it 2 mins before doing this bit of data entry - the elves are working on it" to avoid or prevent the human/bot race condition.

Also, happy if some of these more subtle points about how we _ideally_ want this to work need to be punted to an issue in the interest of getting something usable but imperfect done ahead of events next week.

Reply:
As per the command above, I think I've made this a little simpler and consistent.